### ISSUE TITLE
Evaluation of multi-variate Y

### ISSUE THREAD

**Issue by minimrbanana** [stevie1023/cspn-gpspn#3]
Hi Martin,

I have a question regarding the evaluation.
Assume our data has two outputs Y1 and Y2, then how do we evaluate the MAE and NLPD of them
I think for MAE we can take the sum/average, is it? i.e. MAE(Y1)+MAE(Y2)
What about the NLPD? I would say we apply sum(+) at the **product node** and weighted sum at the **sum node**, to propagate the NLPD bottom up to the root. 

Please let me know if my understandings above is wrong.

**Comment by trappmartin** [comment]
Hi, for the MAE I would take the average over Ys. For NLPD I would suggest to use two different approaches.

1. NLPD by propagation through the SPN as you suggest.
2. NLPD of the Gaussian approximation. Stevie gets for each new observation a multivariate Gaussian with dimensionality equal to the number of Ys and you can compute the NLPD simply based on this Gaussian approximation. This is what I did in the paper as it is using the approximation to a GP while 1. is computing the NLPD based on the more flexible SPN which is a mixture of GPs.

Does that answer your question?

**Comment by trappmartin** [comment]
Approach 1. is essentially just evaluating the NLPD at the leaves and propagation of the result bottom up as usual in SPNs. Products are products and sums are weighted sums, but of course all computed in log space. 

I think that’s what you wrote, I’m not entirely sure. :)

**Comment by minimrbanana** [comment]
I got your second approach. 
IMO the second looks better and more straightforward, as it makes use of the multivariate Gaussian directly. Let's try the second approach first!

**Comment by stevie1023** [comment]
Hi Martin, now I still have a couple of questions ahead. 1. according to what proposed in the link you shared for the computation of  sum nodes, I come across some singular covariance matrix which makes the multivariate gaussian distribution inf for that very observation. Is it normal to have singular matrix for some observations or covariance matrix are guaranteed to be positive-definite?
2. When building the structure, I randomly selected scopes of y in product nodes, should I use independence test instead? or random selection will do?
Thanks~

**Comment by trappmartin** [comment]
Hi Stevie,

> 1. according to what proposed in the link you shared for the computation of sum nodes, I come across some singular covariance matrix which makes the multivariate Gaussian distribution inf for that very observation. Is it normal to have a singular matrix for some observations or covariance matrix are guaranteed to be positive-definite?

Yes, this can happen if the noise variance of your likelihood model is too small. I'm aware that you easily get singular covariance matrices if you write your own GP code. But I was expecting that the package you use might take care of it. 

- What are the hyperparameters of your Gaussian process experts? 
- How are those learned?

You might want to add some additional variance to the covariance matrix if necessary. 
Something like this:

```python
Sigma = ... # the covariance matrix
np.fill_diagonal(Sigma, Sigma.diagonal() + 0.01)
```

My python skills are only very rudimentary. You probably know better how to do this. 

> 2. When building the structure, I randomly selected scopes of y in product nodes, should I use independence test instead? or random selection will do?

I can't tell which one is going to work better tbh. But we can evaluate an independence test based structure construction later on and see if it works better or not. Good point!

**Comment by stevie1023** [comment]
Hi Martin, below is my experiment results
![image](https://user-images.githubusercontent.com/51037813/96443895-eccc8c80-123f-11eb-8fa9-1de7d785fec9.png)
As you can see, generally the two models both have worse results when tested with abnormal dataset against normal dataset(which is a right trend). Our model with product nodes have competitive results against models without product nodes w.r.t rmse and mae( which I train three times to have three y), however, NLPD seems much worse( I took sum over three NLPDs for three y output). I am wondering why is this...... My code is also updated to the repo
![image](https://user-images.githubusercontent.com/51037813/96444298-9dd32700-1240-11eb-9241-23cff391f44b.png)for your reference:)

**Comment by trappmartin** [comment]
Hi, what is normal or abnormal? Are those some synthetic datasets?

**Comment by trappmartin** [comment]
Do you compute NLPD for as SPN without product nodes as follows:
sum_j sum_i - log p(y_i,j | x_i)
?

**Comment by stevie1023** [comment]
> Hi, what is normal or abnormal? Are those some synthetic datasets?

Yes, synthetic data from windmill.

**Comment by stevie1023** [comment]
> Do you compute NLPD for as SPN without product nodes as follows:
> sum_j sum_i - log p(y_i,j | x_i)
> ?

Oh no.. I computed as normal distribution

**Comment by trappmartin** [comment]
The conditional probability should be the probability density of the univariate Normal distribution for location x_i. This is Normal distribution is the posterior predictive distribution.

**Comment by trappmartin** [comment]
> > Hi, what is normal or abnormal? Are those some synthetic datasets?
> 
> Yes, synthetic data from windmill.

Could you evaluate it on a real dataset?

**Comment by trappmartin** [comment]
So you compute the following for the NLPD.

For each datum x_i you obtain the posterior predictive distribution. This is a univariate Normal for each Y_i,j or a multivariate Normal in case the SPN has product nodes.

Then you compute the log density of the true outcome, i.e. log p(y_i,j | x_i) or p(y_i | x_i) respectively. 

The NLPD is then a mean of the negated terms. 

With product nodes:
NLPD = - 1/N sum^N_i p(y_i | x_i)

Without product nodes:
NLPD = - 1/N sum^N_i sum_j p(y_i,j | x_i)
           = - 1/N sum_j sum^N_i p(y_i,j | x_i)

**Comment by stevie1023** [comment]
is NLPD = - 1/N sum^N_i sum_j p(y_i,j | x_i)= - 1/N sum_j sum^N_i p(y_i,j | x_i) for one Y_i,j(one channel) or three channels altogether? suppose we have j=3,  the final result should be -1/N*3*(sum^N_i log[p(y_i,1 | x_i)]+(sum^N_i log[p(y_i,2 | x_i)]+(sum^N_i log[p(y_i,3 | x_i)]?

**Comment by stevie1023** [comment]
> > > Hi, what is normal or abnormal? Are those some synthetic datasets?
> > 
> > 
> > Yes, synthetic data from windmill.
> 
> Could you evaluate it on a real dataset?

Yes, I'll evaluate it on the datasets in your DSGM paper

**Comment by trappmartin** [comment]
If you have three output dimensions Y_1, Y_2, Y_3, you will have three independent SPNs (if the SPN has no product nodes). The NLPD is then computed as follows:

NLPD = - 1/N [ sum_i log p_1(y_i,1 | x_i) + log p_2(y_i,2 | x_i) + log p_2(y_i,3 | x_i) ]

where p_1 is the predictive posterior distribution of the SPN for Y_1, p_2 is the predictive posterior distribution of the SPN for Y_2, and so on.

**Comment by trappmartin** [comment]
This essentially a multivariate Gaussian with diagonal covariance matrix. 
See: https://markusthill.github.io/gaussian-distribution-with-a-diagonal-covariance-matrix/

**Comment by stevie1023** [comment]
> If you have three output dimensions Y_1, Y_2, Y_3, you will have three independent SPNs (if the SPN has no product nodes). The NLPD is then computed as follows:
> 
> NLPD = - 1/N [ sum_i log p_1(y_i,1 | x_i) + log p_2(y_i,2 | x_i) + log p_2(y_i,3 | x_i) ]
> 
> where p_1 is the predictive posterior distribution of the SPN for Y_1, p_2 is the predictive posterior distribution of the SPN for Y_2, and so on.

Yes this is what I currently do. But now I am not sure did I compute p(y|x) or only p(y)...

**Comment by trappmartin** [comment]
The posterior predictive distribution at point x is obtained through the steps you implemented. The Gaussian distribution that you construct for every observation / at every point x, is already a conditional distribution. Conditioned on the location x. So you only need to evaluate the log density for the true outcome to compute p(y | x).

Your model, with product nodes, should consistently perform better than one without product nodes. Or if there is no dependence between the output dimensions, then they should perform equally well. If this is not the case, you might have gotten unlucky with the structure or you might have a bug. :)

**Comment by trappmartin** [comment]
One more thing, you should only add some value to the diagonal of the covariance matrix if it’s singular. Also try to add only a very small value. Adding a larger value, e.g. 0.01, will increase the NLPD artificially.

**Comment by stevie1023** [comment]
> If you have three output dimensions Y_1, Y_2, Y_3, you will have three independent SPNs (if the SPN has no product nodes). The NLPD is then computed as follows:
> 
> NLPD = - 1/N [ sum_i log p_1(y_i,1 | x_i) + log p_2(y_i,2 | x_i) + log p_2(y_i,3 | x_i) ]
> 
> where p_1 is the predictive posterior distribution of the SPN for Y_1, p_2 is the predictive posterior distribution of the SPN for Y_2, and so on.

Yes this is what I currently do. But now I am not sure did I compute p(y|x) or only p(y)...

**Comment by stevie1023** [comment]
> The posterior predictive distribution at point x is obtained through the steps you implemented. The Gaussian distribution that you construct for every observation / at every point x, is already a conditional distribution. Conditioned on the location x. So you only need to evaluate the log density for the true outcome to compute p(y | x).
> 
> Your model, with product nodes, should consistently perform better than one without product nodes. Or if there is no dependence between the output dimensions, then they should perform equally well. If this is not the case, you might have gotten unlucky with the structure or you might have a bug. :)

I cannot find any more bug..This is driving me crazy.:) I'll try with another real dataset and see how that works.Thanks for your explanation!

**Comment by trappmartin** [comment]
Could you explain how you compute the covariance matrix now. I don’t seem to understand https://github.com/stevie1023/cspn-gpspn/blob/f583cb05eacb110147bc95bb18e40dce08457e7e/prod_inference.py#L44. 😅

**Comment by minimrbanana** [comment]
> > > Hi, what is normal or abnormal? Are those some synthetic datasets?
> > 
> > 
> > Yes, synthetic data from windmill.
> 
> Could you evaluate it on a real dataset?

Hi martin, I asked Stevie to try with this synthetic dataset, also want to see if our model can give larger NLPD to ood samples. We will also try the multivariate datasets from your paper.

**Comment by trappmartin** [comment]
Ok. I don’t mind a synthetic dataset. Sorry for intervening. The only thing to make sure is that the Ys are not independent. :)

**Comment by minimrbanana** [comment]
> If you have three output dimensions Y_1, Y_2, Y_3, you will have three independent SPNs (if the SPN has no product nodes). The NLPD is then computed as follows:
> 
> NLPD = - 1/N [ sum_i log p_1(y_i,1 | x_i) + log p_2(y_i,2 | x_i) + log p_2(y_i,3 | x_i) ]
> 
> where p_1 is the predictive posterior distribution of the SPN for Y_1, p_2 is the predictive posterior distribution of the SPN for Y_2, and so on.

I think I have also explained this part to Stevie. Please correct me if the following is wrong.
1) for the with-prod structure, we have already p(y_i|x_i) at test time, so we only need to take the mean of that for all samples. 
2) for the no-prod case, we only have p(y_i_j|x_i) at test time. since it is in the log-space, we just sum up over join order to have p(y_i|x_i). 
*) Now the only question is if the calculation of NLPD is correct or not.

**Comment by trappmartin** [comment]
Yes, that's correct.

**Comment by minimrbanana** [comment]
Two observations from the table are:
1) Assume the calculation NLPD is fine, the NLPD of SPN with smaller GP-leaf threshold(e.g. 500) is lower. That I think is because with smaller training data, the GP provides smaller variance for test samples. Do you think this understanding is valid?
2) The NLPD of no-prod case is smaller, which is not expected. I agree, there must be something wrong.

**Comment by minimrbanana** [comment]
> Could you explain how you compute the covariance matrix now. I don’t seem to understand
> 
> https://github.com/stevie1023/cspn-gpspn/blob/f583cb05eacb110147bc95bb18e40dce08457e7e/prod_inference.py#L44
> 
> . 😅

this is her dirty code, I think it is from the link (https://math.stackexchange.com/questions/147497/covariance-matrix-of-mixture-distribution) for the 3 terms in the last equation.

**Comment by trappmartin** [comment]
1. I would expect the NLPD to behave similar to the RMSE. So larger threshold should also result in better estimates of the posterior predictive distribution. A smaller threshold would create more discontinuities and could, potentially, result in worse generalisation and higher NLPD. All of this really depends on the hyperparameters. So best would be to consistently use the same hyperparameters (for the GPs) in all setups and all models.

**Comment by trappmartin** [comment]
Hm, I think this looks slightly off. But I'm probably not understanding the code. haha..

1. I would expect `c3` to be `mu_x * mu_x .transpose(...)` but it seems it's something else.
2. `col1` and `col2` looks correct.

**Comment by minimrbanana** [comment]
> 1. I would expect the NLPD to behave similar to the RMSE. So larger threshold should also result in better estimates of the posterior predictive distribution. A smaller threshold would create more discontinuities and could, potentially, result in worse generalisation and higher NLPD. All of this really depends on the hyperparameters. So best would be to consistently use the same hyperparameters (for the GPs) in all setups and all models.

this is also one of my questions. In our experiments how to choose the splitting threshold(the max number of samples in a GP leaf)? should we set a fixed one for all datasets, or let it be data-dependent, e.g. 10% of #data ?

**Comment by trappmartin** [comment]
I think you can choose it based on the marginal log likelihood. I had a fixed threshold but that's likely not a good strategy.

So I would suggest computing the marginal log likelihood for a range of thresholds and select the one with the highest marginal log likelihood. This is similar to a Bayes factor based model selection, I think.

**Comment by minimrbanana** [comment]
> Hm, I think this looks slightly off. But I'm probably not understanding the code. haha..
> 
> 1. I would expect `c3` to be `mu_x * mu_x .transpose(...)` but it seems it's something else.
> 2. `col1` and `col2` looks correct.

seems you are right 😅

**Comment by minimrbanana** [comment]
btw, regarding the noise on the diagonal hack, I think, "theoretically" if we apply the mixture of cov matrix as described from the link, it should keep PSD, right? 
And the 0.01 on the diagonal you suggested, should be add to the final cov matrix which is obtained from bottom up propagation, is it?

**Comment by trappmartin** [comment]
Yes it should theoretically be fine without but numerics are a devil.

Yes, it should only be applied at the end.

**Comment by stevie1023** [comment]
> Could you explain how you compute the covariance matrix now. I don’t seem to understand
> 
> https://github.com/stevie1023/cspn-gpspn/blob/f583cb05eacb110147bc95bb18e40dce08457e7e/prod_inference.py#L44
> 
> . 😅
SORRY we had a power cut yesterday and I was forced to stop working:) [I hate Chinese dormitories....

**Comment by stevie1023** [comment]
> Hm, I think this looks slightly off. But I'm probably not understanding the code. haha..
> 
> 1. I would expect `c3` to be `mu_x * mu_x .transpose(...)` but it seems it's something else.
> 2. `col1` and `col2` looks correct.

I'm so sorry about my awful mathematical ability... Now I corrected that and gonna update the results soon!

**Comment by stevie1023** [comment]
> 1. I would expect the NLPD to behave similar to the RMSE. So larger threshold should also result in better estimates of the posterior predictive distribution. A smaller threshold would create more discontinuities and could, potentially, result in worse generalisation and higher NLPD. All of this really depends on the hyperparameters. So best would be to consistently use the same hyperparameters (for the GPs) in all setups and all models.

Yes ,now I keep all the hyperparameters the same for the trials

**Comment by stevie1023** [comment]
Hi, first big apology for the mistake in my code which resulted in such a big mess...... feeel so sorry for my carelessness and being unfamiliar with all the mathematical knowledge...I'll work on that for sureee!
![image](https://user-images.githubusercontent.com/51037813/96535513-905f8080-12c4-11eb-8883-ca34bda4fd81.png) 
This is the updated result which seems reasonable to me. But from this dataset the advantages with product nodes is not obvious, and now we have more points to focus on,1. setting threshold for leave size based on mll 2. independence test which splitting output space(Y), right?

**Comment by minimrbanana** [comment]
Hi Stevie, I think we can see the advantage of prod-nodes. The NLPDs of with-prod-case are all lower, which fulfils our assumption(lower the better, with same threshold). 
Regarding the leaf size setting, I would suggest to add one more line below NLPE, being the mLL, just to see the mLL of the 3 thresholds. 
Also we need results from real world datasets. let me edit the Google docs.

**Comment by minimrbanana** [comment]
> I think you can choose it based on the marginal log likelihood. I had a fixed threshold but that's likely not a good strategy.
> 
> So I would suggest computing the marginal log likelihood for a range of thresholds and select the one with the highest marginal log likelihood. This is similar to a Bayes factor based model selection, I think.

hi Martin, for this mLL, as we have sum nodes, we have to calculate it after training, and then propagate the mLL from each leaf bottom up, is it? Otherwise, what do we do with the mLLs of the same data from multiple leaves?

**Comment by trappmartin** [comment]
I would have to do the derivations first myself too. Intuitively I'd say yes, but let's do the math first. I'm somewhat busy today but will be able to respond in more details tomorrow.

Stevie: Do you update the sum node weight in your current implementation? If not, you should initialise them uniformly, e.g. if you have two children the weights are `[0.5, 0.5]`, and then update them through posterior inference. See algorithm 1 in http://proceedings.mlr.press/v38/peharz15.pdf. Or alternatively you can have a look at my implementation here: 
https://github.com/trappmartin/DeepStructuredMixtures/blob/bf43a03ee5f78ed8dda903cd4dd88651531dde4b/src/common.jl#L323-L332

 Computing the NLPD should always be done using the updated weights.

**Comment by stevie1023** [comment]
> I would have to do the derivations first myself too. Intuitively I'd say yes, but let's do the math first. I'm somewhat busy today but will be able to respond in more details tomorrow.
> 
> Stevie: Do you update the sum node weight in your current implementation? If not, you should initialise them uniformly, e.g. if you have two children the weights are `[0.5, 0.5]`, and then update them through posterior inference. See algorithm 1 in http://proceedings.mlr.press/v38/peharz15.pdf. Or alternatively you can have a look at my implementation here:
> https://github.com/trappmartin/DeepStructuredMixtures/blob/bf43a03ee5f78ed8dda903cd4dd88651531dde4b/src/common.jl#L323-L332
> 
> Computing the NLPD should always be done using the updated weights.

Yes, the current implementation is based on the updated weights(based on mll during training)

**Comment by trappmartin** [comment]
Actually, I don’t need to do the derivation. It’s anyways in the paper already. Because we optimise the MLL. So yes, the MLL is just the mll for each leaf propagated up.

Stevie: Sounds great. Do you update the weights after each iteration or only at the end?

**Comment by stevie1023** [comment]
> Actually, I don’t need to do the derivation. It’s anyways in the paper already. Because we optimise the MLL. So yes, the MLL is just the mll for each leaf propagated up.
> 
> Stevie: Sounds great. Do you update the weights after each iteration or only at the end?

Only at the end- -

**Comment by trappmartin** [comment]
Perfect.

**Comment by stevie1023** [comment]
> > Actually, I don’t need to do the derivation. It’s anyways in the paper already. Because we optimise the MLL. So yes, the MLL is just the mll for each leaf propagated up.
> > Stevie: Sounds great. Do you update the weights after each iteration or only at the end?
> 
> Only at the end- -

During GP optimization, the weights are not relevant,right? I update the weights once I finished training for that GP leave.

**Comment by stevie1023** [comment]
> Perfect.

But my implementation so far for MLL is only a scaler number instead of a vector for every GP leaf. So I take down MLLs for every leaf, and take the average of that?

**Comment by trappmartin** [comment]
What do you mean? The marginal log likelihood should be computed as in my AISTATS paper. See hyper parameter optimisation.

**Comment by stevie1023** [comment]
> What do you mean? The marginal log likelihood should be computed as in my AISTATS paper. See hyper parameter optimisation.

Yes I saw the computations, but I used packages from GPytorch which saved me all the lengthy calculations. For each GP leave we should obtain a set of mlls whose size should equal to observations falling into that covariate space,but the package only returns the mean of all the mlls in that leaf.  So I have no idea how to do with that...:)

**Comment by trappmartin** [comment]
Hm, you should obtain the marginal log likelihood (see Eq. 9 in https://arxiv.org/pdf/1910.04536.pdf) for each leaf node. This should be a scalar for the whole dataset located at the respective leaf node, which is of course a subset of the whole dataset. So you can simply call the GPytorch function and propagate the result upwards.

**Comment by minimrbanana** [comment]
> Hm, you should obtain the marginal log likelihood (see Eq. 9 in https://arxiv.org/pdf/1910.04536.pdf) for each leaf node. This should be a scalar for the whole dataset located at the respective leaf node, which is of course a subset of the whole dataset. So you can simply call the GPytorch function and propagate the result upwards.

I also thought about this. However, if we propagate upwards this mLL from leaf, should we consider also the #samples at leaf? Otherwise, a GP leaf with many samples will contribute equally as the one with a few samples in the final mLL. What do you think, or this fact does not matter at all?

**Comment by trappmartin** [comment]
We are interested in computing the log of the marginal likelihood of the DSMGP. See:
![grafik](https://user-images.githubusercontent.com/7974003/96706959-8b1f3600-1397-11eb-8ced-ba5c5f3d20f9.png)

There should be no need to renormalise something.

**Comment by trappmartin** [comment]
The thetas are the hyperparameters that we aim to marginalise out.

**Comment by trappmartin** [comment]
Wait, let me rethink this... I have a feeling this is not quite correct...

**Comment by trappmartin** [comment]
Sorry for the confusion. It should be correct as `p(\mathcal{T}_k)` is the prior probability of the kth model in the DSMGP.

**Comment by minimrbanana** [comment]
Hi Martin, thanks for the above equation. I have one question, do Y_(L) and X_(L) represent one sample from training data that belong to this leaf, or a set of samples that belong to this leaf?

**Comment by trappmartin** [comment]
Y_(L) and X_(L) are respectively the covariates and targets that fall into the sub-space leaf L is modelling. So, it's a sub-set of the whole dataset.

**Comment by trappmartin** [comment]
Hope that helps.

**Comment by stevie1023** [comment]
Hi Martin， I also got one question..
![image](https://user-images.githubusercontent.com/51037813/96731999-75872c00-13ea-11eb-9c24-7812148bc49d.png)when I do propogation, what is actually this prior probability p(\mathcal{T}_k)? It should be different to different GP leaves？

**Comment by trappmartin** [comment]
It's a uniform prior. Under this prior, every model is equally likely.

**Comment by stevie1023** [comment]
Hi Martin, I had some result under the parkinson and usflight dataset in your paper, as well as MLL.
for parkinsons:
![image](https://user-images.githubusercontent.com/51037813/96858079-60fe6e80-1492-11eb-8ca6-44b38b3076f0.png)

for usflight, unfortunately the GPU on the machine is too limited to run the whole dataset, so I used the first 10% training dataset to see how that works( I wrongly set the test set to 5w too...)
![image](https://user-images.githubusercontent.com/51037813/96858026-504df880-1492-11eb-9627-b9d6b639037e.png)

Not sure if it is comparative enough if I cannot run the whole dataset, but the result so far looks promising! 
For MLL, I am a little bit confused cuz in parkin dataset, the trend is not expected. However, in windmill dataset, I can see larger MLL with smaller NLPD:
![image](https://user-images.githubusercontent.com/51037813/96858688-306b0480-1493-11eb-8156-f9b9d75448b4.png)

What do you think is the problem?

**Comment by trappmartin** [comment]
> For MLL, I am a little bit confused cuz in parkin dataset, the trend is not expected. However, in windmill dataset, I can see larger MLL with smaller NLPD:

I don't think that the log marginal likelihood necessarily has to behave similar to the NLPD on the test set. Ideally it should, but it doesn't have to be the case as one is computed on the training set and the other on the test set.

The results look quite interesting, though.

> for usflight, unfortunately the GPU on the machine is too limited to run the whole dataset, so I used the first 10% training dataset to see how that works( I wrongly set the test set to 5w too...)

You might need to have more splits at each slit node to get this working.

Also, the log marginal likelihood seems to always linearly increase with the number of observations per expert. Is that correct? If so this will likely not be very helpful to select the number of observations.

**Comment by stevie1023** [comment]
> > For MLL, I am a little bit confused cuz in parkin dataset, the trend is not expected. However, in windmill dataset, I can see larger MLL with smaller NLPD:
> 
> I don't think that the log marginal likelihood necessarily has to behave similar to the NLPD on the test set. Ideally it should, but it doesn't have to be the case as one is computed on the training set and the other on the test set.
> 
> The results look quite interesting, though.
> 
> > for usflight, unfortunately the GPU on the machine is too limited to run the whole dataset, so I used the first 10% training dataset to see how that works( I wrongly set the test set to 5w too...)
> 
> You might need to have more splits at each slit node to get this working.
> 
> Also, the log marginal likelihood seems to always linearly increase with the number of observations per expert. Is that correct? If so this will likely not be very helpful to select the number of observations.

I guess if we have more gp experts(less observations per expert), the absolute value  of mll will increase. It seems that for windmill, the majority of GP experts have a positive mll after training 100 iterations, which makes the final mlls positive;
 for parkin, however, we have more negative mlls after the same training process, so the final mlls are all negative... But I am not sure if this makes sense..

**Comment by trappmartin** [comment]
Yes, that can be. The log marginal likelihood is computed (in our case) based on the density. And the density of a Gaussian is positive unbounded, so the log of it can get positive. If it is the case, this means the noise in the data is low and the noise parameter is also chosen to be small.

**Comment by trappmartin** [comment]
Related question: When you learn the hyper parameters, do you have the same hyper parameters for all experts?

**Comment by trappmartin** [comment]
I to select the number of observations per expert we might actually have to do something like AIC, where we penalise the model complexity, i.e. the complexity of the GPs. But we can think about that later.

**Comment by trappmartin** [comment]
Also, for hyper parameter learning you could also do an approach that is more flexible than tying the hyper parameters. I have a more flexible approach in the paper, but this is pretty slow tbh. Alternatively we could have an independent set of hyper parameters for each NLE model. That would be quite interesting to try.

**Comment by stevie1023** [comment]
> Related question: When you learn the hyper parameters, do you have the same hyper parameters for all experts?

Yes, all the hyperparameters are the same.

**Comment by stevie1023** [comment]
> Also, for hyper parameter learning you could also do an approach that is more flexible than tying the hyper parameters. I have a more flexible approach in the paper, but this is pretty slow tbh. Alternatively we could have an independent set of hyper parameters for each NLE model. That would be quite interesting to try.

Do you mean try different hyperparameters on the models?[confused face```]

**Comment by trappmartin** [comment]
> Do you mean try different hyperparameters on the models?

Ignore what I wrote regarding this for now. It would be a way to make the model more flexible, but it's not crucial. It's more important to see how well it works for some fixed hyperparameters and some fixed number of observations per expert.

**Comment by stevie1023** [comment]
> > Do you mean try different hyperparameters on the models?
> 
> Ignore what I wrote regarding this for now. It would be a way to make the model more flexible, but it's not crucial. It's more important to see how well it works for some fixed hyperparameters and some fixed number of observations per expert.

OK I see.. So mll may not be relevant when choosing leaf size?

**Comment by trappmartin** [comment]
It's relevant, but not on its own as it does not penalise models with a large number of observations per expert. I did not think about that at first.

**Comment by stevie1023** [comment]
> It's relevant, but not on its own as it does not penalise models with a large number of observations per expert. I did not think about that at first.

I thought the prior p(\mathcal{T}_k)  should be different for different GP expert. But no clear idea- -

**Comment by trappmartin** [comment]
The prior can be different if you have some knowledge about the data distribution. A uniform prior is essentially an uninformed choice. When computing predictions, you use the posterior of p(𝒯_k) instead. It might also make sense to use the posterior for model comparison, but this will still not penalise the model complexity.

Does that make sense to you?

We should maybe make a Skype meeting sometime to discuss some of the open questions.

**Comment by minimrbanana** [comment]
Hi Martin, 
I think till now I can almost explain everything to Stevie, except for the mll stuff :(
I think we also need to discuss some details of experimental settings. e.g. for the us flight dataset, too large tree structure leads to out of memory. we are curious how you set it.
Tomorrow, Stevie and I will first try to find a way of meeting(Skype, zoom, teams,....) that works smoothly(due to the GFW), and then let's set up a meeting

**Comment by trappmartin** [comment]
Sounds good!

**Comment by minimrbanana** [comment]
Hi Martin, It seems Skype works well! I just created a Skype group(not sure if you are truly "invited"), let me know if you don't receive anything. Stevie can chat before CET16:00(summer time) everyday and I should be free mostly. I think you can suggest a time and we can try a Skype call.

**Comment by trappmartin** [comment]
Do you have my Skype name? If not, it's real-hoshy.

What about next Monday (26.10.2020) at 11:00 am?

**Comment by stevie1023** [comment]
> Do you have my Skype name? If not, it's real-hoshy.
> 
> What about next Monday (26.10.2020) at 11:00 am?

The time is ok for me~

**Comment by minimrbanana** [comment]
it seems I added the wrong user, it is working now?
btw the time is also good for me

**Comment by trappmartin** [comment]
Cool, then let’s call it a date. :)

**Comment by minimrbanana** [comment]
Hint: because of the winter time, it should be 18:00 for Stevie

Best,
Zhongjie

**Comment by minimrbanana** [comment]
Hello martin, 
I have a question regarding the optimisation of GP leaves.
In our case, there are two parts of the parameters to be optimised, one for the kernel (gpytorch.models.ExactGP) and one for the noise (gpytorch.likelihoods.GaussianLikelihood), right? 
But I missed the conclusion we had in our last discussion, which parameter should be applied for the whole input space?

**Comment by trappmartin** [comment]
Hi,
The conclusion was as follows:

For each output dimension we learn hyper parameters for the likelihood model (the noise parameter) and the kernel function (the length scale parameters).
To do so, we extract global hyper parameters for each output dimension and then globally optimise those.

Say if we have a regression problem that is R^3 -> R^2 then we have 2 noise parameters. The first one is shared across all GP leaves that predict Y1 and the second one is shared across all GP leaves that predict Y2.
In addition to this, we have 3*2 hyper parameters for the kernel functions (considering that we use an ARD kernel). Those hyper parameters are again shared across the respective GP leaves.

We can update the global hyper parameters by aggregating the gradients. So this should be relatively little change in the current code.

**Comment by minimrbanana** [comment]
Instead of "aggregating the gradients", do you think it will be the same, if we just sum up the loss of e.g. Y1 from all the leaves, and optimize the sum at once?

**Comment by minimrbanana** [comment]
```python
# for each output dimension, first initialise the models
self.likelihood = GaussianLikelihood()
self.model = ExactGPModel()
self.optimizer = Adam()
self.model.train()
for i in range(steps): # outer loop for optimiser
    tree_loss = 0
    for j, gp in enumerate(gps): # loop for all leaves in the tree
        tree_loss += -mll(output, self.y) # calculate and sum up the loss from each leaf 
    self.optimizer.step()
```
this is the pseudo code of the above understanding, what do you think?

**Comment by trappmartin** [comment]
> Instead of "aggregating the gradients", do you think it will be the same, if we just sum up the loss of e.g. Y1 from all the leaves, and optimize the sum at once?

Yes. See the log marginal likelihood in my paper. This is the equation we aim to optimise.

**Comment by trappmartin** [comment]
> ```python
> ```python
> # for each output dimension, first initialise the models
> self.likelihood = GaussianLikelihood()
> self.model = ExactGPModel()
> self.optimizer = Adam()
> self.model.train()
> for i in range(steps): # outer loop for optimiser
>     tree_loss = 0
>     for j, gp in enumerate(gps): # loop for all leaves in the tree
>         tree_loss += -mll(output, self.y) # calculate and sum up the loss from each leaf 
>     self.optimizer.step()
> ```
> 
> 
> this is the pseudo code of the above understanding, what do you think?
> ```

Looks good. We just have to make sure that gps contains only the gps for the same output dimension, e.g. Y1.

**Comment by trappmartin** [comment]
Oh it’s written in the comment. I didn’t see that. So yeah, perfect!

**Comment by minimrbanana** [comment]
Nice! thanks, now we are synchronised!
